include/ginkgo/core/factorization/ilu.hpp

```diff
@@ -0,0 +1,153 @@
+
+
+#include <memory>
+
+
+#include <ginkgo/core/base/composition.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+
+
+namespace gko {
+/**
+ * @brief The Factorization namespace.
+ *
+ * @ingroup factor
+ */
+namespace factorization {
+
+
+/**
+ * Represents an incomplete LU factorization -- ILU(0) -- of a sparse matrix.
+ *
+ * More specifically, it consists of a lower unitriangular factor $L$ and
+ * an upper triangular factor $U$ with sparsity pattern
+ * $\mathcal S(L + U)$ = $\mathcal S(A)$
+ * fulfilling $LU = A$ at every non-zero location of $A$.
+ *
+ * @tparam ValueType  Type of the values of all matrices used in this class
+ * @tparam IndexType  Type of the indices of all matrices used in this class
+ *
+ * @ingroup factor
+ * @ingroup LinOp
+ */
+template <typename ValueType = gko::default_precision,
+          typename IndexType = gko::int32>
+class Ilu : public Composition<ValueType> {
+public:
+    using value_type = ValueType;
+    using index_type = IndexType;
+    using matrix_type = matrix::Csr<ValueType, IndexType>;
+
+    std::shared_ptr<const matrix_type> get_l_factor() const
+    {
+        // Can be `static_cast` since the type is guaranteed in this class
+        return std::static_pointer_cast<const matrix_type>(
+            this->get_operators()[0]);
+    }
+
+    std::shared_ptr<const matrix_type> get_u_factor() const
+    {
+        // Can be `static_cast` since the type is guaranteed in this class
+        return std::static_pointer_cast<const matrix_type>(
+            this->get_operators()[1]);
+    }
+
+    // Remove the possibility of calling `create`, which was enabled by
+    // `Composition`
+    template <typename... Args>
+    static std::unique_ptr<Composition<ValueType>> create(Args &&... args) =
+        delete;
+
+    GKO_CREATE_FACTORY_PARAMETERS(parameters, Factory)
+    {
+        /**
+         * Strategy which will be used by the L matrix. The default value
+         * `nullptr` will result in the strategy `classical`.
+         */
+        std::shared_ptr<typename matrix_type::strategy_type>
+            GKO_FACTORY_PARAMETER(l_strategy, nullptr);
+
+        /**
+         * Strategy which will be used by the U matrix. The default value
+         * `nullptr` will result in the strategy `classical`.
+         */
+        std::shared_ptr<typename matrix_type::strategy_type>
+            GKO_FACTORY_PARAMETER(u_strategy, nullptr);
+    };
+    GKO_ENABLE_LIN_OP_FACTORY(Ilu, parameters, Factory);
+    GKO_ENABLE_BUILD_METHOD(Factory);
+
+protected:
+    Ilu(const Factory *factory, std::shared_ptr<const gko::LinOp> system_matrix)
+        : Composition<ValueType>{factory->get_executor()},
+          parameters_{factory->get_parameters()}
+    {
+        if (parameters_.l_strategy == nullptr) {
+            parameters_.l_strategy =
+                std::make_shared<typename matrix_type::classical>();
```

Comment:
Is `automatical` a better default ?

Reply:
I think we still can't use `automatical` on a reference executor when we did not compile a CUDA executor.